### DISCUSSION TITLE
Can firestore js sdk onSnapshot listeners work over an http proxy?

### DISCUSSION THREAD

**Discussion by ZYinMD** [firebase/firebase-js-sdk#6135] [General]
I use the firestore js sdk in my web app, and faced a problem: some of my users are behind a firewall that blocks firebase urls (`https://firebase.googleapis.com/`, `https://firestore.googleapis.com/`, etc ) that the sdk talks to.

I plan to solve it by adding a service worker to intercept all requests to these urls, and forward them to my own server, which will then pass the requests to the firebase urls, then pass back the responses.

I imagine it could work, but I'm not sure if onSnapshot listeners will work too. I don't know the mechanism of the onSnapshot listeners, but I imagine WebSockets or gRPC are involved, but when I look at the devTools, I didn't see any WebSockets or gRPC traffic when I trigger the listener.

In any case, I want to know if my service worker + reverse proxy approach would work. Thanks!

**Comment by trullock** [comment]
Did you get an answer on this?

**Comment by dconeybe** [comment]
I can only speak to Firestore, but in theory, an HTTP proxy should work just fine. You don't need to use the complicated Service Worker setup though, as the Firestore SDK can be configured to connect to a specific host instead of `firestore.googleapis.com` by using the `host` and `ssl` attributes of `FirestoreSettings`: https://firebase.google.com/docs/reference/js/firestore_.firestoresettings.md#firestoresettingshost

**Comment by ZYinMD** [comment]
@trullock my original question was submitted in the context where my firebase web app couldn't be used by Chinese users because they couldn't visit firebase urls. 

The question was submitted 3+ years ago, so I don't know what's the current api. In my 3yo firebase 9.x codebase, this worked, to some extent:
```js
const db = initializeFirestore(app, {
	host: myReverseProxyUrl, // I thought if the proxy can only do http, then onSnapshot listeners must use long poll (?) because by default it uses websocket (?)
	experimentalAutoDetectLongPolling: true, // I think this is true by default nowadays
	// experimentalForceLongPolling: true, // maybe set to true if you're sure you'll always need to use an http proxy
});
```
But I did end up using a service worker, because it's more powerful. I needed to reroute more things than just firestore, for instance "callable functions". (Nowadays I don't know if cloud functions are still in fashion, because ideally you should use a meta framework to put your front and backend code under the same package.json so they can share typings)

I ended up using service worker to reroute `https://firebase.googleapis.com` `https://firestore.googleapis.com` `https://identitytoolkit.googleapis.com` `https://securetoken.googleapis.com` `https://url-to-my-functions.cloudfunctions.net` to my proxy.

**Comment by trullock** [comment]
Thanks for your responses.

I'm actaully trying to instrument Reads and was hoping to intercept http/ws traffic to insert some logging.

I can't see how onSnapshot actually fetches data, I'm getting results but I don't see any WS traffic via chrome debug tools...

I'll have a poke around the clues above. Thanks very much

**Comment by dconeybe** [comment]
FYI Firstore does _not_ use websockets for onSnapshot. Instead, it uses a Google-proprietary protocol called "WebChannel" (https://google.github.io/closure-library/api/goog.net.WebChannel.html).

**Comment by trullock** [comment]
What are WebChannels underneath then, just long polling?

**Comment by dconeybe** [comment]
Long polling is _part of_ it. The low-level details of WebChannel are out of scope for the Firebase repository (and my personal knowledge) but Copilot may be able to answer your low-level questions about the protocol. Here's an example conversation I just tested out with Copilot by clicking the "Chat With Copilot" button at https://github.com/google/closure-library:

https://gist.github.com/dconeybe/2d9d46e36f1b17768147d2d7e35e5634

**Reply by trullock** [reply]
Thanks, will read some more
